pkg/frontend/querymiddleware/retry_test.go

```diff
@@ -73,14 +88,25 @@ func TestRetry(t *testing.T) {
 	} {
 		t.Run(tc.name, func(t *testing.T) {
 			try.Store(0)
-			h := newRetryMiddleware(log.NewNopLogger(), 5, nil).Wrap(tc.handler)
+			var cl countingLogger
+			h := newRetryMiddleware(&cl, 5, nil).Wrap(tc.handler)
 			resp, err := h.Do(context.Background(), nil)
 			require.Equal(t, tc.err, err)
 			require.Equal(t, tc.resp, resp)
+			require.Equal(t, tc.retr, cl.count)
 		})
 	}
 }
 
+type countingLogger struct {
```

Comment:
nit: Perhaps it's overkill for this, but my preference would be to add a new unexported method that returns the number of retries and assert based on that instead of something implicit like number of log calls.

e.g.

```go
func (r retry) doWithRetries(ctx context.Context, req Request) (Response, int, error) { /* ... */ }

func (r retry) Do(ctx context.Context, req Request) (Response, error) { 
    resp, _, err := r.doWithRetries(ctx, req)
    return resp, err
}
```

Reply:
Because this is used via the `querymiddleware.Handler` interface, calling into a different function requires a type assertion.

I noticed there is already a `metrics` which is counting retries, however checking the state of a Prometheus histogram is also ugly.

So, I put the metrics behind an interface and mocked that.  See what you think.